source/server/worker_impl.cc

```diff
@@ -127,15 +127,15 @@ void WorkerImpl::stopListener(Network::ListenerConfig& listener, std::function<v
   });
 }
 
-void WorkerImpl::threadRoutine(GuardDog& guard_dog) {
+void WorkerImpl::threadRoutine(GuardDog& guard_dog, const Event::OptDispatcherStartCb& cb) {
   ENVOY_LOG(debug, "worker entering dispatch loop");
   // The watch dog must be created after the dispatcher starts running and has post events flushed,
   // as this is when TLS stat scopes start working.
   dispatcher_->post([this, &guard_dog]() {
     watch_dog_ = guard_dog.createWatchDog(api_.threadFactory().currentThreadId(),
                                           dispatcher_->name(), *dispatcher_);
   });
-  dispatcher_->run(Event::Dispatcher::RunType::Block);
+  dispatcher_->run(Event::Dispatcher::RunType::Block, cb);
```

Comment:
The proposed changes to WorkerImpl seem unavoidable.  See the previous comment about using post instead of additional args to dispatcher run; the cb you're adding could be called by the watchdog create cb in the previous lines.